tests/ci.js

```diff
@@ -1,181 +1,243 @@
+    const runResult = execElmTest([args], false);
+    assertTestSuccess(runResult);
+  }).timeout(60000);
+
+  it('Should fail for failing tests', () => {
+    const args = path.join('tests', '*Fail*.elm');
+    const runResult = execElmTest([args], false);
+    assertTestFailure(runResult);
+  }).timeout(60000);
+});
 
-shell.cd('example-package');
+describe('Testing elm-test on example-application-src', () => {
+  before(() => {
+    shell.pushd('example-application-src');
+  });
 
-assertTestSuccess(path.join('tests', '*Pass*.elm'));
-assertTestFailure(path.join('tests', '*Fail*.elm'));
-assertTestFailure();
-assertTestSuccess('src');
+  after(() => {
+    shell.popd();
+  });
 
-shell.cd('../');
+  it('Should pass successfully', () => {
+    const runResult = execElmTest(['src'], false);
+    assertTestSuccess(runResult);
+  }).timeout(60000);
+});
 
-shell.echo('\n### Testing elm-test on example-application-no-tests');
+describe('Testing elm-test on an application with no tests', () => {
+  before(() => {
+    shell.pushd('example-application-no-tests');
+  });
 
-shell.cd('example-application-no-tests');
+  after(() => {
+    shell.popd();
+  });
 
-assertTestFailure();
+  it('Should fail due to missing tests', () => {
+    const runResult = execElmTest();
+    assertTestFailure(runResult);
+  }).timeout(60000);
+});
 
-shell.cd('../');
+/* ci tests on single elm files */
+describe('Testing elm-test on single Elm files', () => {
+  before(() => {
+    shell.pushd(fixturesDir);
+  });
+
+  after(() => {
+    shell.popd();
+  });
+
+  // passing tests
+  const passingTestFiles = [
+    'Dependency.elm',
+    'One.elm',
+    'Several.elm',
+    'Unexposed.elm',
+  ];
+
+  for (const testToRun of passingTestFiles) {
+    it(`Should succeed for the passing test: ${testToRun}`, () => {
+      const itsPath = path.join('tests', 'Passing', testToRun);
+      const runResult = execElmTest([itsPath]);
+      assertTestSuccess(runResult);
+    }).timeout(5000);
+  }
 
-shell.cd(fixturesDir);
+  it(`Should run every file in tests/Passing`, () => {
+    const filesFound = fs.readdirSync('tests/Passing/');
+    assert.deepStrictEqual(filesFound, passingTestFiles);
+  });
+
+  // failing tests
+  const failingTestFiles = [
+    'Fuzz.elm',
+    'One.elm',
+    'OneRuntimeException.elm',
+    'OneTodo.elm',
+    'Several.elm',
+    'SeveralTodos.elm',
+    'SeveralWithComments.elm',
+    'SplitSocketMessage.elm',
+  ];
+
+  for (const testToRun of failingTestFiles) {
+    it(`Should fail for the failing test: ${testToRun}`, () => {
+      const itsPath = path.join('tests', 'Failing', testToRun);
+      const runResult = execElmTest([itsPath]);
+      assertTestFailure(runResult);
+    }).timeout(5000);
+  }
 
-/* ci tests on single elm files */
+  it(`Should run every file in tests/Failing`, () => {
+    const filesFound = fs.readdirSync('tests/Failing/');
+    assert.deepStrictEqual(filesFound, failingTestFiles);
```

Comment:
```suggestion
    filesFound.sort();
    assert.deepStrictEqual(filesFound, failingTestFiles);
```

I do not think `readDirSync` guarantees the order if the values it returns.

Reply:
Why are we asserting what files are in the fixtures folders?

Reply:
It was a suggestion by @harrysarson to avoid accidentally not running tests : 

> and if somehow we mess around with the directory structure so that tests are not found by shells.ls(...) then we get an error rather than silently not running the tests.

This sounds sensible to me, but I don’t have strong opinions either way.

Reply:
Yeah, I am thinking it would be very easy to add a file to the test fixtures directory and just assume it is tested. If this test fails we get an error message which should act as a fairly good pointer to the array you have to add your test case name to.